### ISSUE TITLE
Paginate Records

### ISSUE THREAD

**Issue by imolloy** [IBM-Cloud/sql-query-clients#21]
I would like to be able to paginate the records, returning only records within a certain range. Currently the `get_result(jobid)` method in the Python interface will attempt to load the entire results file, which may be very large.

**Comment by torsstei** [comment]
@imolloy We did implement a simple pagination in the generic SQL cloud function, which will return the result set by value (i.e. in case of pagination, the individual page) in the response body. See description of the function here: https://hub.docker.com/r/ibmfunctions/sqlquery/

That function is in fact implemented in Python itself on top of ibmcloudsql package. You can take a look how that simple pagination is done here: https://github.com/IBM-Cloud/sql-query-clients/blob/master/Python/cloud_function/sqlfunction.py

**Comment by imolloy** [comment]
Hi @torsstei. This seems to pull the entire file, and then uses the pandas DataFrame to select the range rows. I was thinking about _only_ pulling the desired records. I have a working POC (that still needs a little work), and relies on a `monotonically_increasing_id()` column with a `Range=` option in `get_object` method (guesses byte range to pull and then corrects). I should be able to submit a PR in the coming week if you're interested.

**Comment by torsstei** [comment]
Thank you @imolloy. Yes, that is correct. The benefit of that simplistic pagination mechanism is in fact only there when done in the cloud function, because it only returns the pages from the cloud function to the client (while internally in IBM cloud, between cloud function and SQL Query the full file is always pulled).

Here is another important point. SQL Query is in fact very soon going to support a PARTITION BY option in the STORED AS clause, which will allow to distribute the result across multiple objects. I guess this can be a good foundation for a pagination mechanism in the ibmcloudsql client package. There could still be value with a client side pagination mechanism as you propose and we will process your PR. But I want to make sure you are aware of the full (upcoming) option space.

**Comment by mvle** [comment]
@torsstei One of the motivations for having pagination is to have guaranteed performance on the client side (i.e., won't pull too much data at one time and timeout for instance). Not sure if that is an issue currently when a large file is pulled internally between SQL Query and Cloud Function. 

I think having the PARTITION BY option is great. However, I'm not sure that will guarantee that a single partition won't be too big to pull all at once. Seems like you will still want to have a client side pagination to enforce some QoS.

**Comment by torsstei** [comment]
@mvle Indeed, performance of clients consuming is the key reason for partitioning support. We are going to add a bucketing in addition to the PARTITION BY support, that will allow to specify a custom number of buckets to be created.
